File: java/src/IPv6Tester.java
// package com.ittysensor.ipv6tools;

import java.net.Inet6Address;
import java.net.InetSocketAddress;
import java.net.ServerSocket;
import java.net.Socket;
import java.io.*;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.RejectedExecutionException;
import java.net.NetworkInterface;
import java.net.InetAddress;
import java.util.Collections;
import java.util.List;

public class IPv6Tester {
    private static final int DEFAULT_PORT = 8080;
    private static final String DEFAULT_IPV6_ADDRESS = "::1";
    private static final DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");
    private static final int MAX_CLIENTS = 10;
    private static final ExecutorService executorService = Executors.newFixedThreadPool(MAX_CLIENTS);

    public static void main(String[] args) {
        // Prefer IPv6 addresses
        // System.setProperty("java.net.preferIPv4Stack", "false");
        // System.setProperty("java.net.preferIPv6Addresses", "true");

        if (args.length < 1 || args.length > 3) {
            printUsage();
            System.exit(1);
        }

        String mode = args[0];
        String ipv6Address = args.length > 1 ? args[1] : DEFAULT_IPV6_ADDRESS;
        int port = args.length > 2 ? parsePort(args[2]) : DEFAULT_PORT;

        if (!mode.equals("server") && !mode.equals("client")) {
            printUsage();
            System.exit(1);
        }

        try {
            if (mode.equals("server")) {
                runServer(ipv6Address, port);
            } else {
                runClient(ipv6Address, port);
            }
        } catch (IOException e) {
            System.err.println("Error: " + e.getMessage());
            e.printStackTrace();
        }
    }

    private static void printUsage() {
        System.out.println("Usage: java IPv6Tester <server|client> [ipv6_address] [port]");
        System.out.println("  server|client    - Required. Run as server or client");
        System.out.println("  ipv6_address     - Optional. IPv6 address (default: ::1)");
        System.out.println("  port             - Optional. Port number (default: 8080)");
        System.out.println("\nAvailable IPv6 addresses on this host:");
        printAvailableIPv6Addresses();
        System.out.println("\nJava IPv6 properties:");
        System.out.println("  java.net.preferIPv4Stack: " + System.getProperty("java.net.preferIPv4Stack", "false"));
        System.out.println("  java.net.preferIPv6Addresses: " + System.getProperty("java.net.preferIPv6Addresses", "false"));
        System.out.println("\nExamples:");
        System.out.println("  java IPv6Tester server");
        System.out.println("  java IPv6Tester server 2001:db8:1234:5678::1");
        System.out.println("  java IPv6Tester client 2001:db8:1234:5678::1 8888");
    }

    private static void printAvailableIPv6Addresses() {
        try {
            List<NetworkInterface> interfaces = Collections.list(NetworkInterface.getNetworkInterfaces());
            for (NetworkInterface iface : interfaces) {
                if (iface.isUp() && !iface.isLoopback() && !iface.isVirtual()) {
                    List<InetAddress> addresses = Collections.list(iface.getInetAddresses());
                    for (InetAddress addr : addresses) {
                        if (addr instanceof Inet6Address) {
                            Inet6Address ipv6Addr = (Inet6Address) addr;
                            System.out.printf("  %s: %s\n", iface.getDisplayName(), ipv6Addr.getHostAddress());
                        }
                    }
                }
            }
        } catch (IOException e) {
            System.err.println("Error getting network interfaces: " + e.getMessage());
        }
    }

    private static int parsePort(String portStr) {
        try {
            int port = Integer.parseInt(portStr);
            if (port < 1 || port > 65535) {
                System.err.println("Error: Port must be between 1 and 65535");
                System.exit(1);
            }
            return port;
        } catch (NumberFormatException e) {
            System.err.println("Error: Invalid port number");
            System.exit(1);
        }
        return DEFAULT_PORT; // Will never reach here due to System.exit
    }

    private static void runServer(String ipv6Address, int port) throws IOException {
        try (ServerSocket serverSocket = new ServerSocket()) {
            // Bind to specified IPv6 address
            serverSocket.bind(new InetSocketAddress(ipv6Address, port));
            System.out.println("IPv6 Server started on [" + ipv6Address + "]:" + port);
            System.out.println("Maximum number of simultaneous clients: " + MAX_CLIENTS);

            while (true) {
                try {
                    Socket clientSocket = serverSocket.accept();
                    String clientAddress = clientSocket.getInetAddress().getHostAddress();
                    System.out.println("Client connected from: [" + clientAddress + "]");

                    try {
                        // Handle each client in a separate thread
                        executorService.submit(() -> handleClient(clientSocket, ipv6Address));
                    } catch (RejectedExecutionException e) {
                        System.out.println("Maximum number of clients reached. Rejecting connection from: [" + clientAddress + "]");
                        clientSocket.close();
                    }
                } catch (IOException e) {
                    System.err.println("Error accepting client connection: " + e.getMessage());
                    e.printStackTrace();
                }
            }
        }
    }

    private static void handleClient(Socket clientSocket, String serverAddress) {
        String clientAddress = clientSocket.getInetAddress().getHostAddress();
        try (clientSocket;
             PrintWriter out = new PrintWriter(clientSocket.getOutputStream(), true);
             BufferedReader in = new BufferedReader(new InputStreamReader(clientSocket.getInputStream()))) {
            while (true) {
                // Read client message
                String message = in.readLine();
                if (message == null) {
                    
                    System.out.println("Client disconnected: [" + clientAddress + "]");
                    break;
                }
                System.out.println("Received from client [" + clientAddress + "]: " + message);

                // Send response with timestamp
                String response = "Server received your message at " + LocalDateTime.now().format(formatter) + " at address " + serverAddress;
                out.println(response);

                // Add a delay of 1 second
                Thread.sleep(1000);
            }
        } catch (IOException e) {
            System.err.println("Error handling client [" + clientAddress + "]: " + e.getMessage());
            e.printStackTrace();
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            System.err.println("Sleep interrupted for client [" + clientAddress + "]: " + e.getMessage());
        }
    }

    private static void runClient(String ipv6Address, int port) throws IOException {
        try (Socket socket = new Socket()) {
            // Connect to specified IPv6 address
            socket.connect(new InetSocketAddress(ipv6Address, port));
            System.out.println("Connected to server at [" + ipv6Address + "]:" + port);

            try (PrintWriter out = new PrintWriter(socket.getOutputStream(), true);
                 BufferedReader in = new BufferedReader(new InputStreamReader(socket.getInputStream()))) {

                for (int i = 0; i < 20; i++) {
                    // Send message to server
                    String message = "Hello from IPv6 client at " + LocalDateTime.now().format(formatter);
                    out.println(message);
                    System.out.println("Sent to server: " + message);

                    // Read server response
                    String response = in.readLine();
                    System.out.println("Server response: " + response);

                    // Wait 1 second before next iteration
                    if (i < 19) {
                        Thread.sleep(1000);
                    }
                }
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                System.err.println("Sleep interrupted: " + e.getMessage());
            }
        }
    }
} 

File: python/src/ipv6_tester.py
#!/usr/bin/env python3
import asyncio
import socket
import sys
import datetime
import argparse
from typing import List, Tuple
import logging

class IPv6Tester:
    DEFAULT_PORT = 8080
    DEFAULT_IPV6_ADDRESS = "::1"
    MAX_CLIENTS = 10
    DATE_FORMAT = "%Y-%m-%d %H:%M:%S"

    def __init__(self):
        self.logger = logging.getLogger(__name__)
        self.logger.setLevel(logging.INFO)
        handler = logging.StreamHandler()
        formatter = logging.Formatter('%(message)s')
        handler.setFormatter(formatter)
        self.logger.addHandler(handler)

    def print_usage(self) -> None:
        """Print usage information and available IPv6 addresses."""
        self.logger.info("Usage: python ipv6_tester.py <server|client> [ipv6_address] [port]")
        self.logger.info("  server|client    - Required. Run as server or client")
        self.logger.info("  ipv6_address     - Optional. IPv6 address (default: ::1)")
        self.logger.info("  port             - Optional. Port number (default: 8080)")
        
        self.logger.info("\nAvailable IPv6 addresses on this host:")
        self.print_available_ipv6_addresses()
        
        self.logger.info("\nPython IPv6 related properties:")
        self.logger.info(f"  socket.AF_INET6: {socket.AF_INET6}")
        self.logger.info(f"  socket.has_ipv6: {socket.has_ipv6}")
        
        self.logger.info("\nExamples:")
        self.logger.info("  python ipv6_tester.py server")
        self.logger.info("  python ipv6_tester.py server 2001:db8:1234:5678::1")
        self.logger.info("  python ipv6_tester.py client 2001:db8:1234:5678::1 8888")

    def print_available_ipv6_addresses(self) -> None:
        """Print all available IPv6 addresses on the system."""
        try:
            for interface in socket.getaddrinfo(host=socket.gethostname(), port=None, family=socket.AF_INET6):
                addr = interface[4][0]
                if not addr.startswith('::1'):  # Skip loopback
                    self.logger.info(f"  {interface[3]}: {addr}")
        except Exception as e:
            self.logger.error(f"Error getting network interfaces: {e}")

    async def handle_client(self, reader: asyncio.StreamReader, writer: asyncio.StreamWriter, server_address: str) -> None:
        """Handle individual client connections."""
        client_address = writer.get_extra_info('peername')[0]
        self.logger.info(f"Client connected from: [{client_address}]")

        try:
            while True:
                # Read client message
                data = await reader.readline()
                if not data:
                    self.logger.info(f"Client disconnected: [{client_address}]")
                    break

                message = data.decode().strip()
                self.logger.info(f"Received from client [{client_address}]: {message}")

                # Send response with timestamp
                timestamp = datetime.datetime.now().strftime(self.DATE_FORMAT)
                response = f"Server received your message at {timestamp} at address {server_address}\n"
                writer.write(response.encode())
                await writer.drain()

                # Add a delay of 1 second
                await asyncio.sleep(1)

        except Exception as e:
            self.logger.error(f"Error handling client [{client_address}]: {e}")
        finally:
            writer.close()
            await writer.wait_closed()

    async def run_server(self, ipv6_address: str, port: int) -> None:
        """Run the IPv6 server."""
        try:
            server = await asyncio.start_server(
                lambda r, w: self.handle_client(r, w, ipv6_address),
                ipv6_address,
                port,
                family=socket.AF_INET6
            )
            self.logger.info(f"IPv6 Server started on [{ipv6_address}]:{port}")
            self.logger.info(f"Maximum number of simultaneous clients: {self.MAX_CLIENTS}")

            async with server:
                await server.serve_forever()
        except Exception as e:
            self.logger.error(f"Server error: {e}")

    async def run_client(self, ipv6_address: str, port: int) -> None:
        """Run the IPv6 client."""
        try:
            reader, writer = await asyncio.open_connection(
                ipv6_address,
                port,
                family=socket.AF_INET6
            )
            self.logger.info(f"Connected to server at [{ipv6_address}]:{port}")

            try:
                for i in range(20):
                    # Send message to server
                    timestamp = datetime.datetime.now().strftime(self.DATE_FORMAT)
                    message = f"Hello from IPv6 client at {timestamp}\n"
                    writer.write(message.encode())
                    await writer.drain()
                    self.logger.info(f"Sent to server: {message.strip()}")

                    # Read server response
                    response = await reader.readline()
                    self.logger.info(f"Server response: {response.decode().strip()}")

                    # Wait 1 second before next iteration
                    if i < 19:
                        await asyncio.sleep(1)

            finally:
                writer.close()
                await writer.wait_closed()

        except Exception as e:
            self.logger.error(f"Client error: {e}")

    def main(self) -> None:
        """Main entry point for the IPv6 tester."""
        if len(sys.argv) < 2:
            self.print_usage()
            sys.exit(1)

        mode = sys.argv[1]
        ipv6_address = sys.argv[2] if len(sys.argv) > 2 else self.DEFAULT_IPV6_ADDRESS
        port = int(sys.argv[3]) if len(sys.argv) > 3 else self.DEFAULT_PORT

        if mode not in ['server', 'client']:
            self.print_usage()
            sys.exit(1)

        try:
            if mode == 'server':
                asyncio.run(self.run_server(ipv6_address, port))
            else:
                asyncio.run(self.run_client(ipv6_address, port))
        except KeyboardInterrupt:
            self.logger.info("\nShutting down...")
        except Exception as e:
            self.logger.error(f"Error: {e}")
            sys.exit(1)

if __name__ == "__main__":
    tester = IPv6Tester()
    tester.main() 